src/main/java/com/actiontech/dble/config/util/ConfigUtil.java

```diff
@@ -90,170 +88,260 @@ private static ArrayList<String> getBaseNodeSchemasOfDbGroup(String dbGroup, Map
+                        for (PhysicalDbInstance ds : dbGroup.getAllDbInstanceMap().values()) {
+                            if (ds.isDisabled() || !ds.isTestConnSuccess() || ds.isFakeNode()) {
+                                continue;
+                            }
+                            if (ds.getConfig().getDataBaseType() == DataBaseType.MYSQL) {
+                                mysqlDbInstances.add(ds);
+                            } else {
+                                ckDbInstances.add(ds);
+                            }
+                        }
+                    }
                 }
-            });
+            }
 
             List<String> syncKeyVariables = Lists.newArrayList();
-            List<String> mysqlSyncKeyVariables = getMysqlSyncKeyVariables(mysqlDbGroups, needSync, !CollectionUtil.isEmpty(clickHouseDbGroups));
-            Optional.ofNullable(mysqlSyncKeyVariables).ifPresent(syncKeyVariables::addAll);
-            List<String> clickHouseSyncKeyVariables = getClickHouseSyncKeyVariables(clickHouseDbGroups, needSync);
-            Optional.ofNullable(clickHouseSyncKeyVariables).ifPresent(syncKeyVariables::addAll);
+            List<String> clickHouseSyncKeyVariables = getClickHouseSyncKeyVariables(ckDbInstances, false, needSync, !CollectionUtil.isEmpty(mysqlDbInstances));
+            syncKeyVariables.addAll(clickHouseSyncKeyVariables);
+            List<String> mysqlSyncKeyVariables = getMysqlSyncKeyVariables(mysqlDbInstances, false, needSync, !CollectionUtil.isEmpty(ckDbInstances));
+            syncKeyVariables.addAll(mysqlSyncKeyVariables);
             return syncKeyVariables;
         } finally {
             TraceManager.finishSpan(traceObject);
         }
     }
 
-    @Nullable
-    private static List<String> getMysqlSyncKeyVariables(Map<String, PhysicalDbGroup> dbGroups, boolean needSync, boolean existClickHouse) throws InterruptedException, ExecutionException, IOException {
-        String msg = null;
-        List<String> list = new ArrayList<>();
-        if (dbGroups.size() == 0) {
-            //with no dbGroups, do not check the variables
-            return list;
+    private static List<String> getMysqlSyncKeyVariables(Set<PhysicalDbInstance> mysqlDbInstances, boolean isAllChange, boolean needSync, boolean existClickHouse) throws InterruptedException, ExecutionException, IOException {
+        if (mysqlDbInstances.size() == 0)
+            return new ArrayList<>();
+        Map<VariableMapKey, Future<KeyVariables>> keyVariablesTaskMap = new HashMap<>(mysqlDbInstances.size());
+        getAndSyncKeyVariablesForDataSources(mysqlDbInstances, keyVariablesTaskMap, needSync);
+        return diffMysqlKeyVariables(keyVariablesTaskMap, mysqlDbInstances, isAllChange, existClickHouse);
+    }
+
+    private static List<String> getClickHouseSyncKeyVariables(Set<PhysicalDbInstance> ckDbInstances, boolean isAllChange, boolean needSync, boolean existMysql) throws Exception {
```

Comment:
Clickhouse does not support variables, and the select statement should not be issued here theoretically.
fixed in inner-2303